cmd/samFilter/samFilter.go

```diff
@@ -0,0 +1,296 @@
+package main
+
+import (
+	"flag"
+	"fmt"
+	"github.com/vertgenlab/gonomics/bed"
+	"github.com/vertgenlab/gonomics/cigar"
+	"github.com/vertgenlab/gonomics/common"
+	"github.com/vertgenlab/gonomics/convert"
+	"github.com/vertgenlab/gonomics/dna"
+	"github.com/vertgenlab/gonomics/exception"
+	"github.com/vertgenlab/gonomics/fileio"
+	"github.com/vertgenlab/gonomics/numbers"
+	"github.com/vertgenlab/gonomics/sam"
+	"log"
+	"strings"
+)
+
+type Settings struct {
+	InFile           string
+	OutFile          string
+	AlignQualFilter  int
+	AlignLenFilter   int
+	FilterCigar      string
+	CollapseByUmi    bool
+	SingleCellFormat bool
+	FilterByRegion   string
+	FilterByFlag     int
+	SortByPosition   bool
+	OutBam           bool
+	NoHeader         bool
+}
+
+func filterByQuality(a sam.Sam, filter int) bool {
+	if int(a.MapQ) >= filter {
+		return true
+	} else {
+		return false
+	}
+}
+
+func filterByLength(a sam.Sam, filter int) bool {
+	var pass bool = false
+	for _, k := range a.Cigar {
+		if k.Op == 77 && k.RunLength >= filter {
+			pass = true
+		}
+	}
+	return pass
+}
+
+func filterByCigar(a sam.Sam, filter string) bool {
+	if filter == "starrSeqIntrons" {
+		for _, k := range a.Cigar { //filter out introns longer than 500bp (the length of 1 standard STARR-seq construct)
+			if k.Op == 78 && k.RunLength > 500 {
+				return false
+			}
+		}
+	} else {
+		if cigar.ToString(a.Cigar) != filter {
+			return false
+		}
+	}
+	return true
+}
+
+func filterByFlag(a sam.Sam, filter int) bool {
```

Comment:
Not sure what the goal is with this filter. As is you are testing for truth against all flag bits rather than each flag individually like samtools does. If you want to match the samtools functionality you need to do a bitwise test for the specific flag. If that is the goal you can see examples in `sam/flag.go`. It would be something like `return a.Flag & filter == filter`

Reply:
I think I fixed the flag filter based on what you have here, but I am still a bit confused

Reply:
The & operator in Go is shorthand for a bitwise AND function. It tests each bit independently and only returns 1 if both the query and target bit are 1. For example
```golang
Flag 1:      01001001
               AND
Filter Flag: 00001000
              EQUALS
Result:      00001000
```

So if the flag bit you are looking for is true, then the result of AND will always be the same number as the filter flag regardless of what the other bits are set to. If the flag you are looking for is false then the result will always be 0.